1.train_models/log_reg_plate_1_cp_fs_data/nbconverted/train_model.py

```diff
@@ -0,0 +1,237 @@
+    confusion_matrix,
+    ConfusionMatrixDisplay,
+    precision_score,
+    accuracy_score,
+    classification_report,
+)
+from sklearn.model_selection import train_test_split, KFold
+
+import matplotlib.pyplot as plt
+import seaborn as sns
+
+from joblib import dump, load
+
+
+# ## Find the git root Directory
+
+# In[ ]:
+
+
+# Get the current working directory
+cwd = Path.cwd()
+
+if (cwd / ".git").is_dir():
+    root_dir = cwd
+
+else:
+    root_dir = None
+    for parent in cwd.parents:
+        if (parent / ".git").is_dir():
+            root_dir = parent
+            break
+
+# Check if a Git root directory was found
+if root_dir is None:
+    raise FileNotFoundError("No Git root directory found.")
+
+
+# # Seed and Generator for Reproducibility
+
+# In[ ]:
+
+
+rnd_val = 0  # Random value for all seeds
+rng = np.random.default_rng(seed=rnd_val)  # random number generator
+
+
+# ## Create paths
+
+# In[ ]:
+
+
+filename1 = "Plate_1_sc_norm_fs.parquet"
+plate_path = Path(
+    f"{root_dir}/nf1_painting_repo/3.processing_features/data/feature_selected_data"
+)
+
+path1 = plate_path / filename1
+
+data_path = Path("data")
+
+# Create the parent directories if they don't exist
+data_path.mkdir(parents=True, exist_ok=True)
+
+
+# ## Generate plate dataframe
+
+# In[ ]:
+
+
+# Returns the dataframe generated by the csv
+platedf = pd.read_parquet(path1)
+
+
+# # Preprocess Data
+
+# ## Create Classes
+
+# In[ ]:
+
+
+target_column = "Metadata_genotype"
+stratify_column = "Metadata_Well"
+
+# These represent the fractions of the entire dataset
+train_val_frac = 0.85
+test_frac = 1 - train_val_frac
+val_frac = 0.15
+
+
+# ## Down-sample and stratify by well
+
+# In[ ]:
+
+
+smallest_gene = platedf[target_column].value_counts().min()
+platedata = pd.DataFrame()
+
+for gene in platedf[target_column].unique():
+    df = platedf.loc[platedf["Metadata_genotype"] == gene]
+    df_frac = smallest_gene / len(df)
+    stratwell = df.groupby(stratify_column, group_keys=False).apply(
+        lambda x: x.sample(frac=df_frac, random_state=rnd_val)
+    )
+    platedata = pd.concat([platedata, stratwell], axis="rows")
+
+
+# ## Stratified Train-test split
+
+# In[ ]:
+
+
+traindf, testdf = train_test_split(
+    platedata,
+    train_size=train_val_frac,
+    stratify=platedata[[target_column, stratify_column]],
+    shuffle=True,
+    random_state=rnd_val,
+)
+
+
+# ## Encode Labels
+
+# In[ ]:
+
+
+le = LabelEncoder()
+testdf["label"] = le.fit_transform(testdf[target_column].values)
+traindf["label"] = le.transform(traindf[target_column].values)
+
+
+# ## Remove unecessary columns
+
+# In[ ]:
+
+
+# Remove Metadata
+feat_col = [col for col in platedata.columns if "Metadata" not in col]
+
+# Keep the label column
+feat_col.append("label")
+
+traindf = traindf[feat_col]
+testdf = testdf[feat_col]
+
+
+# # Model Training
+
+# In[ ]:
+
+
+# Default number of splits
+num_splits = 5
+
+
+def kcross_val(model, X_data, y_data, splits=num_splits):
```

Comment:
Maybe worth adding this function to a separate `.py` file that you can import?

Reply:
Also, just checking, but do you need to implement this logic? Doesn't sklearn contain functions to assist with this? I see `KFold` below, but what about [`cross_val_score`](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.cross_val_score.html#sklearn.model_selection.cross_val_score)?

[`cross_validate`](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.cross_validate.html#sklearn.model_selection.cross_validate) also comes to mind

Whenever possible, we should not reinvent wheels :)

Reply:
My last comment about this, is that I'm not sure we should be selecting models based on accuracy. I think average precision is probably a better metric - here are others that can work in the predefined sklearn functions: https://scikit-learn.org/stable/modules/model_evaluation.html#scoring-parameter

Reply:
That's good to know, I suspected something like that existed, I've always just used KFold, but this might be a simpler approach.

I'll also use average_precision, it might matter, something that could also be explored.